source/server/admin/prometheus_stats.cc

```diff
@@ -332,8 +363,13 @@ uint64_t PrometheusStatsFormatter::statsAsPrometheus(
   metric_name_count += outputStatType<Stats::TextReadout>(
       response, params, text_readouts, generateTextReadoutOutput, "gauge", custom_namespaces);
 
-  metric_name_count += outputStatType<Stats::ParentHistogram>(
-      response, params, histograms, generateHistogramOutput, "histogram", custom_namespaces);
+  if (params.histogram_buckets_mode_ == Utility::HistogramBucketsMode::Summary) {
+    metric_name_count += outputStatType<Stats::ParentHistogram>(
+        response, params, histograms, generateSummaryOutput, "summary", custom_namespaces);
+  } else {
+    metric_name_count += outputStatType<Stats::ParentHistogram>(
```

Comment:
The switch bit makes sense, and while I think that returning an error on unsupported modes makes sense, unfortunately that would be a behavioral break, since this currently works:
```
~  curl -vv http://127.0.0.1:9901/stats/prometheus?histogram_buckets=detailed >/dev/null
*   Trying 127.0.0.1:9901...

> GET /stats/prometheus?histogram_buckets=detailed HTTP/1.1
> Host: 127.0.0.1:9901
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: text/plain; charset=UTF-8
< cache-control: no-cache, max-age=0
< x-content-type-options: nosniff
< date: Fri, 03 Nov 2023 20:26:24 GMT
< server: envoy
< transfer-encoding: chunked
<
```

Defer to you all in terms of how much of a "break" this really is, since it seems like this was mostly unintentionally supported before.

Reply:
My preference is to fail requests for disjoint/detailed even though the buckets-mode is ignored now. @ggreenway WDYT?

Also I think we need to add the histogram_buckets param to source/server/admin/admin.cc in the block below
```
          makeHandler("/stats/prometheus", "print server stats in prometheus format",
```
and you can just give the 2 legal options for histogram_buckets there, patterning after "/stats" handler declaration. That won't stop someone from sending a request, bypassing the UI, with an invalid bucket-mode so you'll have to check that on entry to the prom handler.

Note that /stats?format=prometheus works and redirects to  /stats/prometheus handling.

Reply:
note this is still pending but we can wait for Greg to weigh in.

Reply:
I'm fine with breaking requests with an invalid bucket type request, even though it's currently ignored.

Reply:
Updated to add a `validateParams`, which seemed like the least invasive way to make this change. I tried a few other approaches, but each seemed to have their own problems... would be curious if I'm missing somthing:

#### Option 1. Make `statsAsPrometheus` return an `absl::StatusOr<uint64_t>`
This seemed like a good way to tie validation and emission of all the bucket mode cases in one place, but causes a lot of churn in tests/unrelated things. Additionally, we'd need to move histogram handling to the top of the function to prevent any writes making it to the response before we error out, which feels really weird.

#### Option 2. Parse `Utility::HistogramBucketsMode` into `Utility::PrometheusHistogramBucketsMode`
This change would still give us the benefit of there only being one code path that operates/validates the provided bucket mode, and each switch can be fully exhaustive/not have empty behavior as exists right now. This would be a bit more plumbing, and I'm not sure is the direction you'd want to go with this in the future.

In any case, this now returns a 400 and an error message on invalid bucket modes:
```
~ > curl -vv http://127.0.0.1:9901/stats/prometheus?histogram_buckets=disjoint
* Uses proxy env variable no_proxy == 'cluster.local'
*   Trying 127.0.0.1:9901...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9901 (#0)
> GET /stats/prometheus?histogram_buckets=disjoint HTTP/1.1
> Host: 127.0.0.1:9901
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< content-type: text/plain; charset=UTF-8
< cache-control: no-cache, max-age=0
< x-content-type-options: nosniff
< date: Mon, 08 Jan 2024 21:20:52 GMT
< server: envoy
< transfer-encoding: chunked
<
* Connection #0 to host 127.0.0.1 left intact
unsupported Prometheus histogram bucket mode
```